doc: move package config docs to separate page

aduh95 opened the pull request:
This part of the docs aims to contain documentation regarding package configuration that covers both ESM and CJS realms.

Refs: https://github.com/nodejs/modules/issues/539

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

aduh95 commented:
@DerekNonGeneric thanks for the input, I agree that is something important to have in mind.

> breaking pre-existing links is not a viable path forward

IIRC, it's considered OK to break things for experimental features, and this PR is focused mostly on topics in the `esm.md` docs (still experimental at the time of writing). So your comment is about the [Addenda: Package manager tips](https://nodejs.org/docs/latest/api/modules.html#modules_addenda_package_manager_tips) and [Core modules](https://nodejs.org/docs/latest/api/modules.html#modules_core_modules) sections of `modules.md`, is that correct?

> Links to the new section locations below headings

Just to be clear, are you referring to having something like `Section moved to [package config page](…).` for those sections? That could work for me.

DerekNonGeneric commented:
> [&hellip;] it's considered OK to break things for experimental features [&hellip;]

That's not entirely true. The APIs that are marked as experimental are subject to change (breaking), but there is no justification to entirely break the documentation page links.

> So your comment is about the Addenda: Package manager tips and Core modules sections of modules.md, is that correct?

These shouldn't be entirely broken links, correct.

> Just to be clear, are you referring to having something like Section moved to [package config page](…). for those sections? That could work for me.

We have done [something like this in the past](https://github.com/nodejs/modules/blob/master/doc/plan-for-new-modules-implementation.md), so it may be seen as more of an acceptable approach to take here. If you're able to find prior art similar to this in the actual API ref docs, that would be even closer to what could be seen as more acceptable.

aduh95 commented:
> 301 redirects

I've done some digging about that, turns out that wouldn't work here: the hash part of the URL (E.G.: #modules_core_modules) is not transmitted to the server, this information is only available at the client side.

An "automatic" redirection would require involving some hacky JS stuff, I doubt that's a road we want to walk.

Anyway, I have added redirection links in my last commit, PTAL.

guybedford commented:
@DerekNonGeneric the modules documentation has not yet stabilized and we reguarlarly break the links still. I do not see why this PR should be treated differently from that perspective. Agreed changing ESM doc links should start to be blocked soon though.

guybedford commented on doc/api/modules_pkg_config.md:
```diff
@@ -0,0 +1,716 @@
+
+## Packages
+
+### Package entry points
```
Before entry points we should include the info about `"type": "module"` and some basic modules v CommonJS questions. I believe moving that section from the ESM docs verbatim should cover it, but it may also need some massaging to make it more general.

aduh95 replied:
I'd like to do that in another PR if that's OK with you. My plan is actually to relive #32970 to document all the `package.json` fields.

guybedford replied:
Agreed this involves more involved refactoring of the content that can be done as a follow-up. As long as that follow-up is on the roadmap then sure.

aduh95 commented on doc/api/modules.md:
```diff
@@ -80,57 +80,8 @@ by checking `require.main.filename`.
-also add the `/usr/lib/node_modules` folder to the `$NODE_PATH` environment
-variable. Since the module lookups using `node_modules` folders are all
-relative, and based on the real path of the files making the calls to
-`require()`, the packages themselves can be anywhere.
```
Should I link to it in the new page or just leave it as is?

guybedford replied:
I'd just leave it as-is. The cycles information is CommonJS-specific. The REPL / bin information is automatic stuff in npm. NODE_PATH isn't supported for ESM I don't believe. If we wanted to restate the node_modules information in the new docs page we can do that separately based on a more modern interpretation. We could possibly cut back on the content here over time too as a kind of "deprecation" maybe?

DerekNonGeneric commented on .eslintrc.js:
```diff
@@ -44,6 +44,7 @@ module.exports = {
       files: [
         'doc/api/esm.md',
         'doc/api/modules.md',
+        'doc/api/modules_pkg_config.md',
```
If there's going to be a separate page for this, is there any way I can convince you to call it `package_json.md`? I would think the desired permalink for this page would be `https://nodejs.org/api/package_json.html` (to maintain consistency w/ using underscores for spaces). Taking npm's lead on the naming here might be wise.

- https://docs.npmjs.com/creating-a-package-json-file
- https://docs.npmjs.com/configuring-npm/package-json.html

aduh95 replied:
>  is there any way I can convince you to call it `package_json.md`

I have no strong feelings regarding the URL. I'll follow your suggestion unless someone objects. Give me a minute.

aduh95 commented:
Based on comments from @DerekNonGeneric and @guybedford, I have removed the changes in `modules.md` from this PR. I have also added HTML anchors and links to the new doc to maintain backward compatibility with links to the current docs. PTAL.

DerekNonGeneric commented:
@aduh95, do you have an opinion on https://github.com/nodejs/node/pull/34748#discussion_r469979503? 👀

DerekNonGeneric commented on doc/api/package_json.md:
```diff
@@ -0,0 +1,648 @@
+# Modules: Package configuration
```
```suggestion
# Modules: The `package.json` file
```

aduh95 replied:
I'd like to have input from other collaborators on this; that's actually the approach I had taken on #32970, but @GeoffreyBooth had some outstanding objections and it never get through.

DerekNonGeneric commented on doc/api/index.md:
```diff
@@ -36,6 +36,7 @@
 * [Inspector](inspector.html)
 * [Internationalization](intl.html)
 * [Modules](modules.html)
+* [Modules: Package configuration](package_json.html)
```
```suggestion
* [Modules: The `package.json` file](package_json.html)
```

DerekNonGeneric commented on doc/api/esm.md:
```diff
@@ -196,664 +196,24 @@ unspecified.
+<a id="esm_approach_2_isolate_state"></a>
+
+This section was moved to
+[Modules: Package configuration](package_json.html#package_json_packages).
```
```suggestion
[Modules: The `package.json` file](package_json.html).
```

Is this URL fragment supposed to go somewhere? I can't find where to.

aduh95 replied:
That's to scroll down below the TOC

DerekNonGeneric replied:
Oh, I guess that's supposed to be a link to the heading at https://github.com/nodejs/node/pull/34748/files#diff-29fe0e955d3b8213fb3d5a0eca4773c1R11?

Would that fragment really link to that heading? I don't see how. 😕

aduh95 replied:
To be precise, it's a link to this h2 heading:

https://github.com/nodejs/node/blob/32f61aeac1ccd8cfff97d9d94cac860381120df2/doc/api/package_json.md#L11

> Would that fragment really link to that heading? I don't see how.

It does on my local branch 🤷‍♂️ I'm not sure where this behavior is defined, but don't worry `make test-doc` does check all the links to make sure none is broken.

DerekNonGeneric commented on doc/api/esm.md:
````diff
@@ -196,664 +196,24 @@ unspecified.
-  }
-}
-```
+<!-- Anchors to make sure old links find a target -->
+<a id="esm_package_entry_points"></a>
+<a id="esm_main_entry_point_export"></a>
+<a id="esm_subpath_exports"></a>
+<a id="esm_package_exports_fallbacks"></a>
+<a id="esm_exports_sugar"></a>
+<a id="esm_conditional_exports"></a>
+<a id="esm_nested_conditions"></a>
+<a id="esm_self_referencing_a_package_using_its_name"></a>
+<a id="esm_internal_package_imports"></a>
+<a id="esm_dual_commonjs_es_module_packages"></a>
+<a id="esm_dual_package_hazard"></a>
+<a id="esm_writing_dual_packages_while_avoiding_or_minimizing_hazards"></a>
+<a id="esm_approach_1_use_an_es_module_wrapper"></a>
+<a id="esm_approach_2_isolate_state"></a>
````
Have you already considered making these into individual clickable links to each section? That would be more like what I had in mind. If this is already causing an inconvenience, I think we should at least make it as seemless as possible.

aduh95 replied:
Please reconsider this, I really think that would set a bad precedent if we forbid ourselves to ever move an experimental documentation.
With the current version, links to those sections that are not set to an exact version would be a bit still supported, while it limits the visual bloat for users of future generations who won't care that the documentation used to be ordered differently once upon a time.

Nonetheless, if you feel strongly about this, I'm open to work on it.

DerekNonGeneric replied:
> Nonetheless, if you feel strongly about this, I'm open to work on it.

I don't really feel strongly about this, but it seems like you may have already considered what I had in mind. I think it would be best to wait and see what others' opinions are on the matter. I think either way would be fine by me.

DerekNonGeneric commented:
I would like to ping a few folks who are involved in the development of the three main package managers who obviously use this file for packagey-type things. My impression is that it would be best to keep this page as agnostic as possible, although perhaps alluding to the fact that `package.json` is shared by various tools would be of some value here too.

Looking gather some opinions if we've achieved that well enough or what some next best steps would be.

/cc @ruyadorno @shellscape @arcanis

Mainly curious to know if there are things that should belong on this page that we should make clear as it would be the first time Node.js would have a page dedicated to this file. Thanks in advance!

DerekNonGeneric commented:
> I like the four-page separation. Perhaps `Package configuration` over `Package authoring` since in theory package configuration applies to non-published packages too. The source maps documentation seems fine to go with the `module` core module page since the API is related to that.
> &mdash; https://github.com/nodejs/modules/issues/539#issuecomment-670113337

Oh wow, sorry @guybedford, I'm just reading this now. I was not aware that other ideas had been proposed. 😬

DerekNonGeneric commented:
I think this page still needs the "experimental API" documentation warning.

DerekNonGeneric commented on doc/api/package_json.md:
```diff
@@ -0,0 +1,648 @@
+# Modules: Package configuration
+
+<!-- type=misc -->
+
```
```suggestion
> Stability: 1 - Experimental

```

arcanis commented on doc/api/package_json.md:
```diff
@@ -0,0 +1,648 @@
+
+### Internal package imports
+
+In addition to the `"exports"` field it is possible to define internal package
+import maps that only apply to import specifiers from within the package itself.
+
+Entries in the imports field must always start with `#` to ensure they are
+clearly disambiguated from package specifiers.
+
+For example, the imports field can be used to gain the benefits of conditional
+exports for internal modules:
```
Not a comment about this PR by itself but ...

Wait, is that really the case? Does that mean that any package doing an import will cause the resolver to have to backtrack the whole filesystem hierarchy just to find the nearest package.json? Why not just let users define their own sub-packages defining `exports` when they wish to do such overrides, since they are in control anyway?

I'm probably missing something, but it seems like a perf-costly syntactic sugar 😕

guybedford replied:
Internal imports must start with `#`, so it's only in those cases.

That said, self resolution does require this package.json resolution to be done. But then so does the "type": "module" check itself, whereby all package scopes will always be memoized as part of the resolution process anyway.

GeoffreyBooth commented:
Elsewhere we had discussed making this page about package authoring in general, not specific to just `package.json`. That way the page is free to include sections that aren't necessarily `package.json` fields.

Could we maybe call this page `Modules: Packages`? And the introduction could explain what a package is, first, before explaining what package scope is.

aduh95 commented:
> Nice - I'd probably just add a section mentioning that this only covers the fields that Node understands, and maybe add a link to the [Yarn](https://yarnpkg.com/configuration/manifest) / [npm](https://docs.npmjs.com/files/package.json) documentations for those that belong to package managers. It might help some less advanced users that would find this page through SEO.

@arcanis right, I mean to do that in a following PR that documents `package.json` fields that are supported by Node.js.

> Could we maybe call this page `Modules: Packages`? And the introduction could explain what a package is, first, before explaining what package scope is.

@GeoffreyBooth That seems like a good tradeoff to me. I'm not sure what difference there is between a package scope and a package though.

GeoffreyBooth commented:
> I'm not sure what difference there is between a package scope and a package though.

A package is what's listed in `package.json` `"dependencies"`, like `"lodash"`. Within a package, i.e. within `./node_modules/lodash`, there is at least one package _scope,_ which is the folder that contains the `package.json` file: `./node_modules/lodash/package.json` and all subfolders _until_ a subfolder with another `package.json` file is found.

So for example if there was `./node_modules/lodash/dist/package.json`, then `./node_modules/lodash/dist` would be a different package scope than `./node_modules/lodash`. It's similar to JavaScript when you might define a function within the body of another function: the “child” function has its own scope inside the parent function's scope.

ruyadorno commented:
I also think it's a good idea to loop in the folks from @nodejs/package-maintenance WG

lholmquist commented on doc/api/package_json.md:
```diff
@@ -0,0 +1,649 @@
+# Modules: Package configuration
+
+<!-- type=misc -->
+> Stability: 1 - Experimental
```
This seems a little weird to me given the title of the document "The package.json file".  I was reading this as the package.json is at the experimental phase

DerekNonGeneric replied:
Understandable, but that's just how it looks in the source code. Compare to the ESM page.

![image](https://user-images.githubusercontent.com/17770407/90566726-f9b4fd00-e176-11ea-87e2-475e2002fc23.png)

lholmquist replied:
Then maybe a change to the title to "Modules: ECMAScript Modules Package Configuration"?   not really tied to that, though

GeoffreyBooth replied:
I was envisioning that we would rename the file `doc/api/modules_packages.md` and title it `Modules: Packages`.

`esm.md` could be likewise renamed `modules_esm.md` and titled `Modules: ECMAScript modules`.

`modules.md` could become `modules_commonjs.md` and titled `Modules: CommonJS modules`.

Work would need to be done to create redirect links for all the old headings.

Regarding stability, only the ECMAScript modules page would have a page-level experimental warning. On the packages page, individual sections should have their own stability levels listed (as not everything related to packages is experimental, such as `"main"` etc.).

arcanis replied:
Frankly I feel like this page describes a concept perpendicular to what the community understand by "packages", and would be clearer if named "Export Configuration", "Exported Modules", or something similar (ie anything that isn't "package").

> Elsewhere we had discussed making this page about package authoring in general, not specific to just `package.json`. That way the page is free to include sections that aren't necessarily `package.json` fields.

@GeoffreyBooth I understand, but then the title and perspective can be modified when it comes to that, if ever. In its current state, I really think it will be confusing for people landing on this page and seeing no reference to dependencies... and given that exports are a power user feature, I find likely that more people will land on this page thinking about regular package management.

aduh95 replied:
> `esm.md` could be likewise renamed `modules_esm.md` and titled `Modules: ECMAScript modules`.
> 
> `modules.md` could become `modules_commonjs.md` and titled `Modules: CommonJS modules`.

That's covered by #34663. We can't rename the pages without breaking existing links so that's been ruled out.

> Work would need to be done to create redirect links for all the old headings.

That'd be ideal, unfortunately I don't have the bandwidth to do that. If someone wants to pick it up, you're very welcome.

> I was envisioning that we would rename the file `doc/api/modules_packages.md` and title it `Modules: Packages`.

The first section is called `Packages` already, that seems a bit redundant, what are your thoughts?

aduh95 replied:
> Frankly I feel like this page describes a concept perpendicular to what the community understand by "packages"

This page aims to cover all the topics that are related to the module system, but are not ESM nor CJS specific. It's doesn't have to be "packages" indeed, I'm definitely open to suggestions.

On a side note, npm defines a package as `a file or directory that is described by a package.json file` ([source](https://docs.npmjs.com/about-packages-and-modules)). That's unfortunate npm and Node.js uses the same terms with very different definitions, and that's a good point for not using it as the title of the document…

guybedford replied:
A package is a folder with a package.json in it. I'm not sure how this is ambigous between any of the definitions?

aduh95 replied:
A package may be a file for npm. But OK let's not get distracted by that.

I've tried to address the comments in my last commit, PTAL.

aduh95 commented:
Rebased on master to solve merge conflict.

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -0,0 +1,660 @@
+
+A folder containing a `package.json` file, and all subfolders below that folder
+until the next folder containing another `package.json` file, are a _package
+scope_.
```
The package scope concept isn’t mentioned elsewhere on this page. Should we move the “Package scope and file extensions” section, and possibly other parts of the ESM enabling section, here? All the content about package scope and `"type"` applies equally to CommonJS as ESM.

DerekNonGeneric replied:
There's a section for terms, right?

```html
<dt>Package scope</dt>
<dd>

folder containing a `package.json` file, and all subfolders below that folder
until the next folder containing another `package.json`

</dd>
```

DerekNonGeneric replied:
Looks like we still have everything over at the following URL.

https://github.com/nodejs/modules/blob/master/doc/terminology.md

Should we add this definition to there?

GeoffreyBooth replied:
> Should we add this definition to there?

That document was written to help guide the modules group in coming up with the new implementation. It isn't part of the Node docs.

The docs need to explain package scope in order to explain `"type"`. So it needs to be in the docs. I'm thinking that the whole Enabling section needs to move to the Packages page, and the intro sentences rewritten from the perspective of “how to enable ESM” to instead be “how a package tells Node whether it’s CommonJS or ESM.” And then the ESM page still needs an Enabling section, something like:

> ## Enabling
>
> Node.js treats JavaScript code as CommonJS modules by default. Authors can tell Node.js to treat JavaScript code as ECMAScript modules via the `.mjs` file extension, the `package.json` `"type"` field, or the `--input-type=module` flag. See [Determining Module System]().

Where the link at the end links to what was the Enabling section, now living on the Packages page. And the above brief paragraph is all we need as a new Enabling section in the ESM page.

aduh95 replied:
Done in 7043be8a2172b4d16ee2b348d51c7197087f1dec, PTAL

aduh95 commented:
@DerekNonGeneric @guybedford @GeoffreyBooth Can you have another look? I'd like to have it landed if it's ready so I can iterate on it with `package.json` fields :)

DerekNonGeneric commented:
Oh, you made each subheading experimental. Thanks, I prefer that actually. Looking at how other pages do this, that's *the way* actually.

GeoffreyBooth commented on doc/api/packages.md:
```diff
@@ -0,0 +1,830 @@
+[the full specifier path]: modules_esm.html#modules_esm_mandatory_file_extensions
+[the dual CommonJS/ES module packages section]: #packages_dual_commonjs_es_module_packages
+[ES modules]: esm.html
+[CommonJS]: modules.html
```
Can you please double-check that the content now in `packages.md` is up-to-date with the latest changes on `master` for `esm.md`? It seems like it is, since I see `--conditions`, but we should verify before we merge this in.

And anyone else on this thread, we should hold other PRs that change `esm.md` until this one lands.

aduh95 replied:
> Can you please double-check that the content now in packages.md is up-to-date with the latest changes on master for esm.md

I'm quite confident I have included all the recent changes: everytime a section on master changes, it creates a conflict on rebase so it's not too difficult to spot where changes have been made.

guybedford commented on doc/api/packages.md:
```diff
@@ -0,0 +1,830 @@
+
+A folder containing a `package.json` file, and all subfolders below that folder
+until the next folder containing another `package.json`, are a
+_package scope_. The `"type"` field defines how to treat `.js` files
+within the package scope. Every package in a
+project’s `node_modules` folder contains its own `package.json` file, so each
+project’s dependencies have their own package scopes. If a `package.json` file
+does not have a `"type"` field, the default `"type"` is `"commonjs"`.
```
Two issues here: (1) it's not strictly correct that everything in node_modules necessarily has a package.json, (2) the type boundary does stop on node_modules which is important information too here.

Suggested rewording:

```suggestion
_package scope_. Package scopes do not carry through `node_modules` folders. The
`"type"` the module format of `.js` files within the package scope. If a
`package.json` file does not have a `"type"` field, the default is `"commonjs"`.
```

guybedford commented on doc/api/esm.md:
```diff
@@ -44,822 +44,41 @@ Node.js contains support for ES Modules based upon the
+details.
+
+<!-- Anchors to make sure old links find a target -->
+<i id="esm_package_entry_points"></i>
+<i id="esm_main_entry_point_export"></i>
+<i id="esm_subpath_exports"></i>
+<i id="esm_package_exports_fallbacks"></i>
+<i id="esm_exports_sugar"></i>
+<i id="esm_conditional_exports"></i>
+<i id="esm_nested_conditions"></i>
+<i id="esm_self_referencing_a_package_using_its_name"></i>
+<i id="esm_internal_package_imports"></i>
+<i id="esm_dual_commonjs_es_module_packages"></i>
+<i id="esm_dual_package_hazard"></i>
+<i id="esm_writing_dual_packages_while_avoiding_or_minimizing_hazards"></i>
+<i id="esm_approach_1_use_an_es_module_wrapper"></i>
+<i id="esm_approach_2_isolate_state"></i>
```
Is there a reason for not just making these `<a name="">` tags?

aduh95 replied:
That's was my original approach, however:

* `name` attribute is deprecated for `<a>` elements in HTML5.
* `<a>` elements get decorated with CSS – even when they are empty. You can actually observe that in https://nodejs.org/docs/latest/api/deprecations.html, when you hover above the `DEPXXXX` heading:
    - ![image](https://user-images.githubusercontent.com/14309773/91336880-e3401e80-e7d2-11ea-8a1f-da65fd3e05b4.png)
    - ![image](https://user-images.githubusercontent.com/14309773/91336924-f18e3a80-e7d2-11ea-894e-9af9254169ad.png)

If `<i>` is not appropriate, I'm fine using something else BTW.

guybedford replied:
Ok was just curious, I'm fine with this approach. I did not know the `"name"` element was deprecated! So `window.esm_conditional_exports` being defined it is :P

guybedford commented on doc/api/packages.md:
```diff
@@ -0,0 +1,830 @@
+
+### Resolving user conditions
+
+When running Node.js, custom user conditions can be added with the
+`--conditions` or `-u` flag:
```
This was just reverted on master, so would be good to get this change in:

```suggestion
When running Node.js, custom user conditions can be added with the
`--conditions` flag:
```

guybedford commented on doc/api/packages.md:
```diff
@@ -0,0 +1,830 @@
+A folder containing a `package.json` file, and all subfolders below that folder
+until the next folder containing another `package.json` file, are a _package
+scope_.
+
```
Can we add another sort of summary here as to what fields are defined here, and a pointer to npm definitions, maybe something like:

```
Node.js uses the following fields in the package.json:

* [`"type"`](#link)
* [`"main"`](#link)
* [`"exports"`](#link)
* [`"imports"`](#link)

Other projects define and use their own fields, such as those used by [npm for defining dependencies](https://docs.npmjs.com/files/package.json).
```

aduh95 replied:
I'd very much prefer to do that in a following PR if that's OK: I promise you that's something I want to do (it's been several months I have started pushing for it 😅), I'm worried nothing will get done if I try to do it in one PR.

guybedford replied:
Ok great, let's do it next of course.

aduh95 commented:
Is this ready to land? If no one objects, I think we can add `Author ready` label now :)

GeoffreyBooth commented:
Let's fix that one sentence but otherwise this looks good to me.

GeoffreyBooth commented:
@isaacs This PR just moves existing content (and adds stubs to link to the new locations). I was thinking we would do follow-up PRs with any _changes_ to that content. Do you mind saving the notes on content for a follow-up?

isaacs commented:
@GeoffreyBooth That makes perfect sense.  If you are going to do that PR, do you want to tag me into it, or ping me once this merges, or are you fine just taking these notes and moving forward at that time?  I don't follow this repo too closely, so I don't want it to block on my attention which might be elsewhere at that time :)

GeoffreyBooth commented:
> @GeoffreyBooth That makes perfect sense. If you are going to do that PR, do you want to tag me into it, or ping me once this merges, or are you fine just taking these notes and moving forward at that time? I don't follow this repo too closely, so I don't want it to block on my attention which might be elsewhere at that time :)

I won't be doing such a PR; it's kind of up to the person who wants to see these changes (i.e. you) to make a PR with the changes you want to see happen. This one looks ready to merge in, so you can fork off of this branch (or off of `master` once this is in) to make such a PR. The notes are good, so maybe if you don't have time, someone else like the indefatigable @aduh95 might feel like running with them in a new PR.

guybedford commented:
I think it should be a condition of this PR landing that follow-up work in line with the points already discussed here will be done by someone though.

aduh95 commented:
> I think it should be a condition of this PR landing that follow-up work in line with the points already discussed here will be done by someone though.

@guybedford Are you saying you'd rather not see this PR get through if nobody volunteers? 🙁 It feels a bit harsh to dismiss my work based on suggestions out of the scope of the PR.

Why not create an issue linking to @isaacs comments, maybe label the issue `Good first issue`, and see if someone picks it up?

Let's not forget: perfect is the enemy of good. If we want to wait until we have created perfect docs to land anything, Geoffrey won't call "indefatigable" for long 😉

isaacs commented:
Go ahead and land this. I'll do a pr with my suggested edits. I mistakenly thought this was new content, so comments were appropriate, but if it's just a content move, it makes more sense to not have edits ride along on the same commit.

isaacs commented:
Ie, to @guybedford's point, this is me volunteering (assuming no one else beats me to it), so that condition is met. Let's align perfect and good here and make them not enemies ;)

guybedford commented:
Thanks @isaacs for volunteering here, I just wanted to ensure there is some follow-through here that's all so things don't get dropped, not trying to be difficult and certainly didn't mean to make work for you either. I'm happy to PR my own previous comments as well further so sounds like we are good to go here then.

DerekNonGeneric commented:
Landed in https://github.com/nodejs/node/commit/dbc5c172cc4064569ef46f21df60fb1129d250bc

aduh95 commented on doc/api/packages.md:
```diff
@@ -0,0 +1,828 @@
+
+A folder containing a `package.json` file, and all subfolders below that folder
+until the next folder containing another `package.json`, are a
+_package scope_. Package scopes do not carry through `node_modules` folders. The
```
> If module.js does `import 'foo/file.js'` is it loaded as cjs or mjs?

That'd be loaded as CJS.
However, `import './node_modules/foo/file.js'` would be loaded as ESM.

MylesBorins commented:
This PR doesn't land cleanly on v14.x. Before we backport I think it would be a good idea for us to ensure all the follow up PRs have been landed so this change gets released in it's best possible form.

Thoughts?

guybedford commented:
Yes, I agree we should rather we wait a little on the release of this docs work until we've completed the subsequent update PRs to a point where the majority of feedback has been incorporated and we are happy with the final state. Specifically landing https://github.com/nodejs/node/pull/34970 and any other follow ups mentioned in this PR.

guybedford commented on doc/api/packages.md:
````diff
@@ -0,0 +1,828 @@
+}
+```
+
+Then any module _in that package_ can reference an export in the package itself:
````
Yes exactly, this name mechanism overrides the existing node_modules lookup using the package boundary as the source of truth.

guybedford commented on doc/api/packages.md:
```diff
@@ -0,0 +1,828 @@
+present. This would likely cause hard-to-troubleshoot bugs.
+
+Aside from writing a stateless package (if JavaScript’s `Math` were a package,
+for example, it would be stateless as all of its methods are static), there are
```
We're referring to internal state specifically here, which is a separate point to the concept of being able to override an interface in JS.

guybedford commented on doc/api/packages.md:
```diff
@@ -0,0 +1,828 @@
+A folder containing a `package.json` file, and all subfolders below that folder
+until the next folder containing another `package.json` file, are a _package
+scope_.
+
```
I believe this sort of outline will be obvious from the generated table of contents since the fields are each their own section.

I'm not sure we should be listing fields that are not defined by Node.js. We do specifically link to the npm package.json definition page though in the package.json fields introduction.

guybedford replied:
I've moved this to the beginning of the reference section itself in the feedback PR.

guybedford commented:
PR for the outstanding feedback here created in https://github.com/nodejs/node/pull/35370.
